src/java.base/share/classes/java/nio/Heap-X-Buffer.java.template

```diff
@@ -278,6 +278,60 @@ class Heap$Type$Buffer$RW$
 
 #if[char]
 
+    //
+    // Use getChars() to load chars directly into the heap buffer array.
+    // For a String or StringBuffer source this improves performance if
+    // a proper subsequence is being appended as copying to a new intermediate
+    // String object is avoided. For a StringBuilder where either a subsequence
+    // or the full sequence of chars is being appended, copying the chars to
+    // an intermedite String in StringBuilder::toString is avoided.
+    //
+    private $Type$Buffer appendChars(CharSequence csq, int start, int end) {
+        checkSession();
+
+        int length = end - start;
+        int pos = position();
+        int lim = limit();
+        int rem = (pos <= lim) ? lim - pos : 0;
+        if (length > rem)
+            throw new BufferOverflowException();
+
+        if (csq instanceof String str) {
+            str.getChars(start, end, hb, ix(pos));
+        } else if (csq instanceof StringBuilder sb) {
+            sb.getChars(start, end, hb, ix(pos));
+        } else if (csq instanceof StringBuffer sb) {
+            sb.getChars(start, end, hb, ix(pos));
+        }
+
+        position(pos + length);
+
+        return this;
+    }
+
+    public $Type$Buffer append(CharSequence csq) {
+#if[rw]
+        if (csq instanceof StringBuilder)
+            return appendChars(csq, 0, csq.length());
```

Comment:
No special casing of String here as append(String) as the implementation in the base class uses put(String) so nothing to be gained.

No special casing of StringBuffer as it is legacy/not-interesting. Also it's caching of the String value means you don't see the difference in micros, is that right?

Reply:
Comments added in https://github.com/openjdk/jdk/pull/13415/commits/c18565eb822aa96a49169c4fffe9ebe203599356.

Reply:
I think this new comment is confusing as it's a reply to a question in the PR rather than to someone reading the code :-)  I think in ab5fe71c without this comment was ready to go.

Reply:
Ah, I thought it was a suggestion. I'll revert it.

Reply:
Reverted in https://github.com/openjdk/jdk/pull/13415/commits/9649549f365ab9d7a76b6ed30cdb03f7c7a6765b.